dsa_test.go

```diff
@@ -0,0 +1,205 @@
+package openssl_test
+
+import (
+	"crypto/dsa"
+	"encoding/asn1"
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/golang-fips/openssl/v2"
+	"github.com/golang-fips/openssl/v2/bbig"
+)
+
+type dsaSignature struct {
+	R, S *big.Int
+}
+
+func TestDSAGenerateParameters(t *testing.T) {
+	var tests = []struct {
+		L, N int
+	}{
+		{1024, 160},
+		{2048, 224},
+		{2048, 256},
+		{3072, 256},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("%d-%d", test.L, test.N), func(t *testing.T) {
+			if openssl.FIPS() {
+				t.Skip("generating DSA parameters with L = 2048 is not supported in FIPS mode")
+			}
+			testGenerateDSAParameters(t, test.L, test.N)
+		})
+	}
+}
+
+func testGenerateDSAParameters(t *testing.T, L, N int) {
+	params, err := openssl.GenerateDSAParameters(L, N)
+	if err != nil {
+		t.Errorf("error generating parameters: %s", err)
+		return
+	}
+
+	P := bbig.Dec(params.P)
+	Q := bbig.Dec(params.Q)
+	G := bbig.Dec(params.G)
+
+	if P.BitLen() != L {
+		t.Errorf("params.BitLen got:%d want:%d", P.BitLen(), L)
+	}
+
+	if Q.BitLen() != N {
+		t.Errorf("q.BitLen got:%d want:%d", Q.BitLen(), L)
+	}
+
+	one := new(big.Int)
+	one.SetInt64(1)
+	pm1 := new(big.Int).Sub(P, one)
+	quo, rem := new(big.Int).DivMod(pm1, Q, new(big.Int))
+	if rem.Sign() != 0 {
+		t.Error("p-1 mod q != 0")
+	}
+	x := new(big.Int).Exp(G, quo, P)
+	if x.Cmp(one) == 0 {
+		t.Error("invalid generator")
+	}
+
+	priv, err := openssl.GenerateKeyDSA(params)
+	if err != nil {
+		t.Errorf("error generating key: %s", err)
+		return
+	}
+
+	testDSASignAndVerify(t, priv)
+}
+
+func testDSASignAndVerify(t *testing.T, priv *openssl.PrivateKeyDSA) {
+	hashed := []byte("testing")
+	sig, err := openssl.SignDSA(priv, hashed[:])
+	if err != nil {
+		t.Errorf("error signing: %s", err)
+		return
+	}
+	pub, err := openssl.NewPublicKeyDSA(priv.DSAParameters, priv.Y)
+	if err != nil {
+		t.Errorf("error getting public key: %s", err)
+		return
+	}
+	if !openssl.VerifyDSA(pub, hashed[:], sig) {
+		t.Error("error verifying")
+		return
+	}
+
+	// Test compatibility with crypto/dsa.
+	priv1 := dsa.PrivateKey{
+		PublicKey: dsa.PublicKey{
+			Parameters: dsa.Parameters{
+				P: bbig.Dec(priv.P),
+				Q: bbig.Dec(priv.Q),
+				G: bbig.Dec(priv.G),
+			},
+			Y: bbig.Dec(priv.Y),
+		},
+		X: bbig.Dec(priv.X),
+	}
+	var esig dsaSignature
+	if _, err := asn1.Unmarshal(sig, &esig); err != nil {
+		t.Error(err)
+		return
+	}
+	if !dsa.Verify(&priv1.PublicKey, hashed[:], esig.R, esig.S) {
+		t.Error("compat: crypto/dsa can't verify OpenSSL signature")
+	}
+	r1, s1, err := dsa.Sign(openssl.RandReader, &priv1, hashed[:])
+	if err != nil {
+		t.Errorf("error signing: %s", err)
+		return
+	}
+	sig, err = asn1.Marshal(dsaSignature{r1, s1})
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	if !openssl.VerifyDSA(pub, hashed[:], sig) {
+		t.Error("compat: OpenSSL can't verify crypto/dsa signature")
+		return
+	}
+}
+
+func TestDSASignAndVerify(t *testing.T) {
```

Comment:
My lack of Go is going to be exposed here :-).  Why the repeated function name with starting with a capital T?

Reply:
It's not a rookie question if you knew that the Go test framework requires test functions to start with the `Test` prefix (with capital T) and have the `(t *testing.T)` signature, else they won't be identified as such and won't be executed when running `go test` (see https://pkg.go.dev/testing). I'll assume you knew that 😄.

So, to your question, `testDSASignAndVerify` (with lowercase T) is just a reusable function that does what says on the tin: test that a given DSA key can be used to sign and verify. Note that this function is called from multiple places, as there are multiple tests that exercise larger workflows. On the other hand, `TestDSASignAndVerify` just exercise the DSA sign/verify workflow using hardcoded inputs, so it is a "small" wrapper around `testDSASignAndVerify`.